ddl/ddl_api.go

```diff
@@ -5118,3 +5122,104 @@ func (d *ddl) AlterIndexVisibility(ctx sessionctx.Context, ident ast.Ident, inde
 	err = d.callHookOnChanged(err)
 	return errors.Trace(err)
 }
+
+func validPlacementSpecs(specs []*ast.PlacementSpec) ([]*placement.Rule, error) {
+	rules := make([]*placement.Rule, 0, len(specs))
+	for k, spec := range specs {
+		rule := &placement.Rule{
+			ID:       "default_rule",
+			Count:    int(spec.Replicas),
+			Override: false,
+		}
+
+		switch spec.Tp {
+		case ast.PlacementAdd:
+		default:
+			return rules, errors.Errorf("invalid placement spec[%d], unknown action type: %d", k, spec.Tp)
+		}
+
+		switch spec.Role {
+		case ast.PlacementRoleFollower:
+			rule.Role = placement.Follower
+		case ast.PlacementRoleLeader:
+			rule.Role = placement.Leader
+		case ast.PlacementRoleLearner:
+			rule.Role = placement.Learner
+		case ast.PlacementRoleVoter:
+			rule.Role = placement.Voter
+		default:
+			return rules, errors.Errorf("invalid placement spec[%d], unknown role: %d", k, spec.Role)
+		}
+
+		for _, label := range strings.Split(spec.Constraints, ",") {
+			var op placement.LabelConstraintOp
+			switch label[0] {
```

Comment:
Is it possible that `label` is empty?

Reply:
Yes, it can be...  Nice catch. Parser only promised that at least one label is valid. There needs more check